File: Tests/OPLFleetAgentCoreTests/PetAssetTests.swift
import CryptoKit
import Foundation
import XCTest

@testable import OPLFleetAgentCore

final class PetAssetTests: XCTestCase {
  func testDiscoversLocalPetAndCachesByMetadata() async throws {
    let root = try makeTemporaryCodexHome()
    defer { try? FileManager.default.removeItem(at: root) }
    let spritesheet = webP(payload: Data("first".utf8))
    let paths = try writePet(in: root, id: "build-fox", spritesheet: spritesheet)
    let originalDate = Date(timeIntervalSince1970: 1_000)
    try FileManager.default.setAttributes(
      [.modificationDate: originalDate], ofItemAtPath: paths.spritesheet.path)
    let catalog = AmbientOpsPetAssetCatalog(codexHome: root)

    let discoveredFirst = await catalog.currentAsset()
    let first = try XCTUnwrap(discoveredFirst)
    XCTAssertEqual(first.definition.id, "build-fox")
    XCTAssertEqual(first.definition.displayName, "Build Fox")
    XCTAssertEqual(first.definition.spriteVersionNumber, 2)
    XCTAssertEqual(first.definition.assetHash, sha256(spritesheet))

    let sameSizeReplacement = webP(payload: Data("other".utf8))
    XCTAssertEqual(sameSizeReplacement.count, spritesheet.count)
    try sameSizeReplacement.write(to: paths.spritesheet)
    try FileManager.default.setAttributes(
      [.modificationDate: originalDate], ofItemAtPath: paths.spritesheet.path)

    let discoveredCached = await catalog.currentAsset()
    let cached = try XCTUnwrap(discoveredCached)
    XCTAssertEqual(cached.definition.assetHash, first.definition.assetHash)

    try FileManager.default.setAttributes(
      [.modificationDate: originalDate.addingTimeInterval(2)],
      ofItemAtPath: paths.spritesheet.path)
    let discoveredUpdated = await catalog.currentAsset()
    let updated = try XCTUnwrap(discoveredUpdated)
    XCTAssertEqual(updated.definition.assetHash, sha256(sameSizeReplacement))
    XCTAssertNotEqual(updated.definition.assetHash, first.definition.assetHash)
  }

  func testRejectsEscapingManifestAndOversizedAsset() async throws {
    let escapingRoot = try makeTemporaryCodexHome()
    defer { try? FileManager.default.removeItem(at: escapingRoot) }
    _ = try writePet(
      in: escapingRoot,
      id: "unsafe",
      spritesheet: webP(payload: Data("x".utf8)),
      spritesheetPath: "../secret.webp"
    )
    let escapingAsset = await AmbientOpsPetAssetCatalog(
      codexHome: escapingRoot
    ).currentAsset()
    XCTAssertNil(escapingAsset)

    let oversizedRoot = try makeTemporaryCodexHome()
    defer { try? FileManager.default.removeItem(at: oversizedRoot) }
    let oversized = Data(
      repeating: 0,
      count: AmbientOpsPetAssetCatalog.maximumAssetBytes + 1)
    _ = try writePet(in: oversizedRoot, id: "too-large", spritesheet: oversized)
    let oversizedAsset = await AmbientOpsPetAssetCatalog(
      codexHome: oversizedRoot
    ).currentAsset()
    XCTAssertNil(oversizedAsset)
  }

  func testIgnoresPetMetadataFieldsOutsideAllowlist() async throws {
    let root = try makeTemporaryCodexHome()
    defer { try? FileManager.default.removeItem(at: root) }
    let directory = root.appendingPathComponent("pets/private-owl", isDirectory: true)
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let manifest = """
      {
        "id": "private-owl",
        "displayName": "Private Owl",
        "spriteVersionNumber": 1,
        "spritesheetPath": "spritesheet.webp",
        "prompt": "never transmit me",
        "response": "nor me",
        "sessionPath": "/Users/private/.codex/sessions"
      }
      """
    try Data(manifest.utf8).write(to: directory.appendingPathComponent("pet.json"))
    try webP(payload: Data("safe".utf8)).write(
      to: directory.appendingPathComponent("spritesheet.webp"))

    let discoveredAsset = await AmbientOpsPetAssetCatalog(codexHome: root).currentAsset()
    let asset = try XCTUnwrap(discoveredAsset)
    let identity = try AmbientOpsMachineIdentity(
      machineID: "private-mac", machineName: "Private Mac", platform: "macOS")
    var tracker = AmbientOpsPetTracker()
    let usage = UsageSnapshot.empty(at: Date(), status: .ready)
    let snapshot = AmbientOpsAgentSnapshot(
      usage: usage,
      identity: identity,
      pet: tracker.snapshot(definition: asset.definition, usage: usage)
    )
    let data = try JSONEncoder().encode(snapshot)
    let payload = try XCTUnwrap(String(data: data, encoding: .utf8))

    XCTAssertFalse(payload.contains("never transmit me"))
    XCTAssertFalse(payload.contains("nor me"))
    XCTAssertFalse(payload.contains("/Users/"))
    XCTAssertFalse(payload.contains("spritesheetPath"))
  }

  private func makeTemporaryCodexHome() throws -> URL {
    let root = FileManager.default.temporaryDirectory
      .appendingPathComponent(UUID().uuidString, isDirectory: true)
    try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
    return root
  }

  @discardableResult
  private func writePet(
    in root: URL,
    id: String,
    spritesheet: Data,
    spritesheetPath: String = "spritesheet.webp"
  ) throws -> (manifest: URL, spritesheet: URL) {
    let directory = root.appendingPathComponent("pets/\(id)", isDirectory: true)
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let manifestURL = directory.appendingPathComponent("pet.json")
    let spritesheetURL = directory.appendingPathComponent("spritesheet.webp")
    let manifest = """
      {
        "id": "\(id)",
        "displayName": "\(id == "build-fox" ? "Build Fox" : id)",
        "spriteVersionNumber": 2,
        "spritesheetPath": "\(spritesheetPath)"
      }
      """
    try Data(manifest.utf8).write(to: manifestURL)
    try spritesheet.write(to: spritesheetURL)
    return (manifestURL, spritesheetURL)
  }

  private func webP(payload: Data) -> Data {
    var data = Data("RIFF".utf8)
    var size = UInt32(payload.count + 4).littleEndian
    data.append(Data(bytes: &size, count: MemoryLayout<UInt32>.size))
    data.append(Data("WEBP".utf8))
    data.append(payload)
    return data
  }

  private func sha256(_ data: Data) -> String {
    SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
  }
}

File: windows/tests/OPLFleetAgent.Core.Tests/PetAssetTests.cs
using System.Buffers.Binary;
using System.Security.Cryptography;
using System.Text;
using OPLFleetAgent.Core;

namespace OPLFleetAgent.Core.Tests;

public sealed class PetAssetTests : IDisposable
{
    private readonly string root = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"));

    [Fact]
    public void DiscoversLocalPetAndCachesByMetadata()
    {
        var firstBytes = WebP("first");
        var spritesheet = WritePet(root, "build-fox", firstBytes);
        var originalDate = DateTime.UnixEpoch.AddSeconds(1_000);
        File.SetLastWriteTimeUtc(spritesheet, originalDate);
        var catalog = new AmbientOpsPetAssetCatalog(root);

        var first = Assert.IsType<AmbientOpsPetAsset>(catalog.CurrentAsset());
        Assert.Equal("build-fox", first.Definition.Id);
        Assert.Equal("Build Fox", first.Definition.DisplayName);
        Assert.Equal(2, first.Definition.SpriteVersionNumber);
        Assert.Equal(Sha256(firstBytes), first.Definition.AssetHash);

        var replacement = WebP("other");
        Assert.Equal(firstBytes.Length, replacement.Length);
        File.WriteAllBytes(spritesheet, replacement);
        File.SetLastWriteTimeUtc(spritesheet, originalDate);
        Assert.Equal(first.Definition.AssetHash, catalog.CurrentAsset()!.Definition.AssetHash);

        File.SetLastWriteTimeUtc(spritesheet, originalDate.AddSeconds(2));
        var updated = Assert.IsType<AmbientOpsPetAsset>(catalog.CurrentAsset());
        Assert.Equal(Sha256(replacement), updated.Definition.AssetHash);
        Assert.NotEqual(first.Definition.AssetHash, updated.Definition.AssetHash);
    }

    [Fact]
    public void RejectsEscapingManifestAndOversizedAsset()
    {
        WritePet(root, "unsafe", WebP("x"), "../secret.webp");
        Assert.Null(new AmbientOpsPetAssetCatalog(root).CurrentAsset());

        var otherRoot = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString("N"));
        try
        {
            WritePet(
                otherRoot,
                "too-large",
                new byte[AmbientOpsPetAssetCatalog.MaximumAssetBytes + 1]);
            Assert.Null(new AmbientOpsPetAssetCatalog(otherRoot).CurrentAsset());
        }
        finally
        {
            Directory.Delete(otherRoot, recursive: true);
        }
    }

    [Fact]
    public void PetManifestCannotAddConversationContentToSnapshot()
    {
        var directory = Path.Combine(root, "pets", "private-owl");
        Directory.CreateDirectory(directory);
        File.WriteAllText(
            Path.Combine(directory, "pet.json"),
            """
            {
              "id": "private-owl",
              "displayName": "Private Owl",
              "spriteVersionNumber": 1,
              "spritesheetPath": "spritesheet.webp",
              "prompt": "never transmit me",
              "response": "nor me",
              "sessionPath": "C:\\Users\\private\\.codex\\sessions"
            }
            """);
        File.WriteAllBytes(Path.Combine(directory, "spritesheet.webp"), WebP("safe"));

        var asset = Assert.IsType<AmbientOpsPetAsset>(
            new AmbientOpsPetAssetCatalog(root).CurrentAsset());
        var usage = UsageSnapshot.Empty(
            DateTimeOffset.FromUnixTimeSeconds(1_000),
            CollectionStatus.Ready);
        var pet = new AmbientOpsPetTracker().Snapshot(asset.Definition, usage);
        var identity = new AmbientOpsMachineIdentity("private-pc", "Private PC", "Windows");
        var json = System.Text.Json.JsonSerializer.Serialize(
            AmbientOpsAgentSnapshot.FromUsage(usage, identity, pet: pet),
            AmbientOpsPushClient.SerializerOptions);

        Assert.DoesNotContain("never transmit me", json, StringComparison.Ordinal);
        Assert.DoesNotContain("nor me", json, StringComparison.Ordinal);
        Assert.DoesNotContain(@"C:\Users", json, StringComparison.Ordinal);
        Assert.DoesNotContain("spritesheetPath", json, StringComparison.OrdinalIgnoreCase);
    }

    public void Dispose()
    {
        if (Directory.Exists(root))
        {
            Directory.Delete(root, recursive: true);
        }
    }

    private static string WritePet(
        string codexHome,
        string id,
        byte[] spritesheet,
        string spritesheetPath = "spritesheet.webp")
    {
        var directory = Path.Combine(codexHome, "pets", id);
        Directory.CreateDirectory(directory);
        File.WriteAllText(
            Path.Combine(directory, "pet.json"),
            $$"""
            {
              "id": "{{id}}",
              "displayName": "{{(id == "build-fox" ? "Build Fox" : id)}}",
              "spriteVersionNumber": 2,
              "spritesheetPath": "{{spritesheetPath}}"
            }
            """);
        var path = Path.Combine(directory, "spritesheet.webp");
        File.WriteAllBytes(path, spritesheet);
        return path;
    }

    internal static void WritePetForUpload(string codexHome) =>
        WritePet(codexHome, "local-pet", WebP("pet pixels"));

    private static byte[] WebP(string payload)
    {
        var payloadBytes = Encoding.UTF8.GetBytes(payload);
        var data = new byte[12 + payloadBytes.Length];
        "RIFF"u8.CopyTo(data);
        BinaryPrimitives.WriteUInt32LittleEndian(data.AsSpan(4), (uint)payloadBytes.Length + 4);
        "WEBP"u8.CopyTo(data.AsSpan(8));
        payloadBytes.CopyTo(data.AsSpan(12));
        return data;
    }

    private static string Sha256(byte[] data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();
}
